smt.go

```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
 	return proof, nil
 }
 
+// ProveClosest generates a SparseMerkleProof of inclusion for the first
+// key with the most common bits as the path provided.
+//
+// This method will follow the path provided until it hits a leaf node and then
+// exit. If the leaf is along the path it will produce an inclusion proof for
+// the key (and return the key-value internal pair) as they share a common
+// prefix. If however, during the tree traversal according to the path, a nil
+// node is encountered, the traversal backsteps and flips the path bit for that
+// depth (ie tries left if it tried right and vice versa). This guarentees that
+// a proof of inclusion is found that has the most common bits with the path
+// provided, biased to the longest common prefix
+func (smt *SMT) ProveClosest(path []byte) (
+	closestPath, closestValueHash []byte, // the closest leaf info for the key provided
+	proof *SparseMerkleProof, // proof of the key-value pair found
+	err error, // the error value encountered
+) {
+	workingPath := make([]byte, len(path))
+	copy(workingPath, path)
+	var siblings []treeNode
+	var sib treeNode
+	var parent treeNode
+	var depthDelta int
+
+	node := smt.tree
+	depth := 0
+	// continuously traverse the tree until we hit a leaf node
+	for depth < smt.depth() {
+		// save current node information as "parent" info
+		if node != nil {
+			parent = node
+		}
+		// resolve current node
+		node, err = smt.resolveLazy(node)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		// if we hit a nil node we backstep to the parent node and flip the path bit
+		// at the parent depth and select the other child
+		if node == nil {
+			node, err = smt.resolveLazy(parent)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			// trim the last sibling node added as it is no longer relevant
+			if len(siblings) > 0 {
+				siblings = siblings[:len(siblings)-1]
+			}
+			depth -= depthDelta
+			// flip the path bit at the parent depth
+			flipPathBit(workingPath, depth)
+		} else {
+			depthDelta = 0
+		}
+		// end traversal when we hit a leaf node
+		if _, ok := node.(*leafNode); ok {
+			break
+		}
+		if ext, ok := node.(*extensionNode); ok {
+			length, match := ext.match(workingPath, depth)
+			if match {
```

Comment:
The thing is we cannot return or continue after this so both need to be in branches of the if/else as the result is we are either at a leaf or inner node afterwards and want to continue to the logic below. Personally I think swapping the shorter branch for the longer branch is purely stylistic here. LMK if you really think it helps to swap the branches around but I dont think it will help readability

Reply:
This is what I had in mind:
```go
if !match {
   node = ext.expand()
} else {
  // large code block
}
```

I personally think it'll help readability, but it is a NIT.

If you strongly disagree, I want push for it.